Intrinsic: introduce minimumnum and maximumnum for IR and SelectionDAG (#96649)
llvm/llvm-project: llvm/docs/LangRef.rst

in particular, if the input is a NaN, then the quiet/signaling bit and payload
are perfectly preserved.

.. _i_fminmax_family:

'``llvm.min.*``' Intrinsics Comparation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Standard:
"""""""""

IEEE754 and ISO C define some min/max operations, and they have some differences
on working with qNaN/sNaN and +0.0/-0.0. Here is the list:

.. list-table::
   :header-rows: 2

   * - ``ISO C``
     - fmin/fmax
     - fmininum/fmaximum
     - fminimum_num/fmaximum_num

   * - ``IEEE754``
     - minNum/maxNum (2008)
     - minimum/maximum (2019)
     - minimumNumber/maximumNumber (2019)

   * - ``+0.0 vs -0.0``
     - either one
     - +0.0 > -0.0
     - +0.0 > -0.0

   * - ``NUM vs sNaN``
     - qNaN, invalid exception
     - qNaN, invalid exception
     - NUM, invalid exception

   * - ``qNaN vs sNaN``
     - qNaN, invalid exception
     - qNaN, invalid exception
     - qNaN, invalid exception

   * - ``NUM vs qNaN``
     - NUM, no exception
     - qNaN, no exception
     - NUM, no exception

LLVM Implementation:
""""""""""""""""""""

> LLVM implements all ISO C flavors as listed in this table.
> Only basic intrinsics list here. The constrained version
> ones may have different behaivor on exception.

Reviewer: ```suggestion
LLVM implements all ISO C flavors as listed in this table, except in the default floating-point environment exceptions are ignored. The constrained versions of the intrinsics respect the exception behavior. 

```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Standard:
"""""""""

IEEE754 and ISO C define some min/max operations, and they have some differences
on working with qNaN/sNaN and +0.0/-0.0. Here is the list:

.. list-table::
   :header-rows: 2

   * - ``ISO C``
     - fmin/fmax
     - fmininum/fmaximum
     - fminimum_num/fmaximum_num

   * - ``IEEE754``
     - minNum/maxNum (2008)
     - minimum/maximum (2019)
     - minimumNumber/maximumNumber (2019)

   * - ``+0.0 vs -0.0``
     - either one
     - +0.0 > -0.0
     - +0.0 > -0.0

   * - ``NUM vs sNaN``
     - qNaN, invalid exception
     - qNaN, invalid exception
     - NUM, invalid exception

   * - ``qNaN vs sNaN``
     - qNaN, invalid exception
     - qNaN, invalid exception
     - qNaN, invalid exception

   * - ``NUM vs qNaN``
     - NUM, no exception
     - qNaN, no exception
     - NUM, no exception

LLVM Implementation:
""""""""""""""""""""

LLVM implements all ISO C flavors as listed in this table.
Only basic intrinsics list here. The constrained version
ones may have different behaivor on exception.

Since some architectures implement minNum/maxNum with +0.0>-0.0,
so we define internal ISD::MINNUM_IEEE and ISD::MAXNUM_IEEE.

> They will be helpful to implement minimumnum/maximumnum.

Reviewer: This description doesn't belong in the LangRef unless they are IR visible. Also the primary difference for these is the signaling nan handling matches the 2008 behavior

Author: OK. Let's drop it.

I mean that we can use
```
fcanonical
fcanonical
MINNUM_IEEE
```
to implement `minimumnum` on such architectures, like ARM64/MIPSr6/LoongArch/PowerPC.

Syntax:
"""""""

This is an overloaded intrinsic. You can use ``llvm.minimumnum`` on any
floating-point or vector of floating-point type. Not all targets support
all types however.

::

      declare float     @llvm.minimumnum.f32(float %Val0, float %Val1)
      declare double    @llvm.minimumnum.f64(double %Val0, double %Val1)
      declare x86_fp80  @llvm.minimumnum.f80(x86_fp80 %Val0, x86_fp80 %Val1)
      declare fp128     @llvm.minimumnum.f128(fp128 %Val0, fp128 %Val1)
      declare ppc_fp128 @llvm.minimumnum.ppcf128(ppc_fp128 %Val0, ppc_fp128 %Val1)

Overview:
"""""""""

The '``llvm.minimumnum.*``' intrinsics return the minimum of the two
arguments, not propagating NaNs and treating -0.0 as less than +0.0.

Arguments:
""""""""""

The arguments and return value are floating-point numbers of the same
type.

Semantics:
""""""""""
If both operands are NaNs (including sNaN), returns qNaN. If one operand
is NaN (including sNaN) and another operand is a number, return the number.
Otherwise returns the lesser of the two arguments. -0.0 is considered to
be less than +0.0 for this intrinsic.

Note that these are the semantics of minimumNumber specified in IEEE 754-2019.

> It has some different with '``llvm.minnum.*``':

Reviewer: ```suggestion
It has some differences with '``llvm.minnum.*``':
```

Syntax:
"""""""

This is an overloaded intrinsic. You can use ``llvm.maximumnum`` on any
floating-point or vector of floating-point type. Not all targets support
all types however.

::

      declare float     @llvm.maximumnum.f32(float %Val0, float %Val1)
      declare double    @llvm.maximumnum.f64(double %Val0, double %Val1)
      declare x86_fp80  @llvm.maximumnum.f80(x86_fp80 %Val0, x86_fp80 %Val1)
      declare fp128     @llvm.maximumnum.f128(fp128 %Val0, fp128 %Val1)
      declare ppc_fp128 @llvm.maximumnum.ppcf128(ppc_fp128 %Val0, ppc_fp128 %Val1)

Overview:
"""""""""

The '``llvm.maximumnum.*``' intrinsics return the maximum of the two
arguments, not propagating NaNs and treating -0.0 as less than +0.0.

Arguments:
""""""""""

The arguments and return value are floating-point numbers of the same
type.

Semantics:
""""""""""
If both operands are NaNs (including sNaN), returns qNaN. If one operand
is NaN (including sNaN) and another operand is a number, return the number.
Otherwise returns the greater of the two arguments. -0.0 is considered to
be less than +0.0 for this intrinsic.

Note that these are the semantics of maximumNumber specified in IEEE 754-2019.

> It has some different with '``llvm.maxnum.*``':

Reviewer: ```suggestion
It has some differences with '``llvm.maxnum.*``':
```
